### PR TITLE
Some `mapreduce` improvement

### PR DESCRIPTION
**Pull request by N5N3**
Seperated from https://github.com/JuliaLang/julia/pull/43604.

That PR wants to make `extrema` rely on our `mapreduce` mechanism to fix `NaN`,`0.0/-0.0` comparison.

I tried my best to keep the git histroy for easier review.

Note: This PR included some test change. I guess CI will keep failing there before we merge this. Is there a better way to handle this problem?

### CODE REVIEW

File: test/higherorderfns.jl
Comment (tkf) [inline test/higherorderfns.jl:730]: Do you mean

```suggestion
# TODO: enable the following once https://github.com/JuliaLang/julia/pull/43604 is merged
```

?

File: test/higherorderfns.jl
Comment (tkf) [inline test/higherorderfns.jl:726]: Does this work after https://github.com/JuliaLang/julia/pull/43604?

```suggestion
    # TODO: enable the following once https://github.com/JuliaLang/julia/pull/43604 is merged
    @test_broken "reducing over an empty" extrema(sparse(z); dims=1)
```

File: test/higherorderfns.jl
Comment (tkf) [inline test/higherorderfns.jl:721]: Does this work after https://github.com/JuliaLang/julia/pull/43604?

```suggestion
    # TODO: enable the following once https://github.com/JuliaLang/julia/pull/43604 is merged
    @test_broken "reducing over an empty" extrema(sparse(C); dims=1) == extrema(C; dims=1)
```

File: test/higherorderfns.jl
Comment (tkf) [inline test/higherorderfns.jl:713]: Does this work after https://github.com/JuliaLang/julia/pull/43604?

```suggestion
    # TODO: enable the following once https://github.com/JuliaLang/julia/pull/43604 is merged
    @test_broken "reducing over an empty" extrema(spzeros(0, 0))
    @test_broken "reducing over an empty" extrema(spzeros(0))
```

File: test/higherorderfns.jl
Comment (tkf) [inline test/higherorderfns.jl:748]: Can we mark these tests as broken so that we don't forget to enable them once #43604 is merged?

```suggestion
function test_extrema(a; dims_test = ((), 1, 2, (1,2), 3))
    for dims in dims_test
        vext = extrema(a; dims)
        vmin, vmax = minimum(a; dims), maximum(a; dims)
        @test_broken all(x -> isequal(x[1], x[2:3]), zip(vext,vmin,vmax))
    end
end
@testset "NaN test for sparse extrema" begin
    for sz = (3, 10, 100)
        A = sprand(sz, sz, 0.3)
        A[rand(1:sz^2,sz)] .= NaN
        test_extrema(A)
        A = sprand(sz*sz, 0.3)
        A[rand(1:sz^2,sz)] .= NaN
        test_extrema(A; dims_test = ((), 1, 2))
    end
end
```

**Review by tkf — COMMENTED** [review]
LGTM. I think it's ready to be merged once it's clarified at what condition the commented out tests can be enabled.

**Comment by N5N3** [comment]
It seems a little bit complex to mark `@test_thrown` as broken. So I just re-comment them.

**Comment by N5N3** [comment]
`Base._mapreduce(f, op, ::IndexCartesian, A::AbstractSparseVector{T})` has been replaced with `Base._mapreduce(f, op, ::IndexLinear, A::AbstractSparseVector{T})`

**Comment by N5N3** [comment]
Just notice that we'd better enable related commits when testing #43604.
So I use `isdefined(Base, :_extrema_rf)` to check whether we want them.

**Comment by N5N3** [comment]
Just notice that we'd better enable related change when testing 43604.
So I change the test to a conditional skip using `isdefined(Base, :_extrema_rf)` to check.

**Review by tkf — APPROVED** [review]
LGTM!

@ViralBShah @KristofferC Can you merge this? (Who has the merge bit?)

**Comment by ViralBShah** [comment]
@tkf I've added this repo to the list of repos that committers can write to. You should be able to merge in the future. Thanks.

**Pull request merged**
